---
title: "Make a COYE BGP map Notebook"
author: "C. M. Bossu"
output:
  html_document:
    df_print: paged
    toc: yes
  html_notebook:
    toc: yes
---

## Overview

To begin with, I first cloned the super cool repository (an RStudio project) at
[https://github.com/eriqande/make-a-BGP-map](https://github.com/eriqande/make-a-BGP-map) to get all the necessary input files, etc.  This clone is for creating a genoscape map for the Willow flycatcher. I then used the base of this code to create a script to generate the genoscape for the Common yellowthroat. This is one small section of the BGP genomic pipeline, with brand new input files specific for Common yellowthroat. Then open up the RStudio project and run though `Make-a-BGP-map-Notebook.Rmd`.

This chronicles the steps taken whilst making a genoscape for the Common yellowthroat
The order in which we discuss building up this map is different from the order in which
we put layers down to actually make the map. Here, we will start with making the
genoscape, which is actually the part that sits atop the whole map.  But we do that
because that is the part that will actually change from species to species, and we 
want everyone to be relatively fresh for understanding those mutable parts.  Once the
genoscape is dealt with, we will then talk about the next layer of country and state
boundaries, and also coastline polygons.  We do this, because it doesn't take too long
to plot these features, and using them is a good way to figure out the desired extent of
your map and to decide upon a projection.  Finally, we will talk about the bottom
layer of the map, which is the raster with earthforms and shading from Natural Earth Data.

## Packages Needed

### Non-standard Packages

This work draws on a few functions that I have in packages that I have up on GitHub,
namely:

1. **Eric's fork** of `tess3r`.  Note that you can't use the default version of `tess3r`,
you have to use Eric's fork of it, which has some extra functionality.
2. Eric's package `genoscapeRtools`

Get those packages like this:
```{r, eval=FALSE}
#remotes::install_github("eriqande/TESS3_encho_sen")  # for special version of tess3r

#remotes::install_github("eriqande/genoscapeRtools")  # for Eric's genoscapeRtools
```

### Standard Packages

The rest of the packages you need can be downloaded from CRAN.  If you don't have
them you should get them: `raster`, `sf`, `fields`, `downloader`, and `tidyverse`. The last one there gets ggplot2
and a number of other packages by Hadley Wickham.

You can get those like this, but comment out once you have them, no need to keep installing these packages
```{r, eval=FALSE}
#install.packages(c("raster", "sf", "tidyverse", "fields", "downloader"))
```

### Load the Packages we will work with

```{r, comment=FALSE, warning=FALSE, message=FALSE}
library(raster)  # important to load before tidyverse, otherwise it masks select()
library(tidyverse)
library(sf)
library(ggspatial)
```

## Making the genoscape

The genoscape can be thought of as the bright colors smeared across space that show where different
genetically identifiable groups of birds reside on the breeding grounds.  These genoscapes
are stored as rasters, and transparency is used to indicate how much confidence one has
in the genetic identification of individuals in different areas.  These rasters are made
by interpolating Q-values from a program like STRUCTURE or ADMIXTURE between individuals
that were sampled in space. 

### Input Data

#### Breeding bird Q-values
We need a matrix of Q-values for individuals.  We have one that we will read in as a
tibble
```{r, message=FALSE}
Q_tibble <- read_table2("~/Dropbox/BGP/COYE/Fluidigm_SNPgenotypes/coye.stateFID.5cCA.Region.rm_miss20.K5.run1.str_locprior.indivq",col_names = T) %>% rename(grp1=Cluster1, grp2=Cluster2, grp3=Cluster3, grp4=Cluster4, grp5=Cluster5)
Q_tibble
```

Column `id` is the sample name and the rest are ancestry fractions to different clusters
(named with three characters) estimated by STRUCTURE.

#### Breeding Bird Lat-Longs
We also need to know where those individuals were sampled in space. We have that here:
```{r, message=FALSE}
met <- read.delim("~/Downloads/SNP Genotyping Master - COYE.txt",header = T,sep="\t",na.strings = NA) %>% filter(State!="NA") %>% 
  dplyr::select(Sample,State,Location,Lat,Long,Year,Month,Day)
met %>% distinct(State)

##Create individual popID
rub<-read_delim("~/Dropbox/BGP/COYE/Fluidigm_SNPgenotypes/COYE.96snp-genotypes.breeding.Rubias.CollectionFID.ped.str.5regionlocation.rubias.txt",delim="\t") %>% rename(Sample=indiv) %>% select(Sample,collection)

#Put the meta data in order of the Qmatrices:
LatLong_tibble <-Q_tibble %>% dplyr::select(Sample) %>% left_join(met) %>% group_by(State) %>% summarise(meanLat=mean(Lat),meanLong=mean(Long))


#LatLong_tibble <- read_tsv("inputs/breeding-bird-lat-longs.tsv")
LatLong_tibble
```

#### Breeding Bird Cluster Colors

The Q-values correspond to different clusters, as shown above.  We must specify
the colors that we wish to assign to each of those clusters.  We do that with a named
vector like this:
```{r}
cluster_colors <-  c(
  grp1 = "#CC0000", #red, CA
  grp2 = "#3399FF", #blue, Midwest
  grp3 = "#9933CC", #purple, Atlantic Coast
  grp4 = "#009933", #green, PNW
  grp5 = "#FF6600")  #orange, SW
```

For fun, we can plot these to see what they look like:
```{r}
enframe(cluster_colors) %>%
  mutate(x = 1:n(),
         y = 1:n()) %>%
  ggplot(aes(x = x, y = y, fill = name)) +
  geom_point(pch = 21, size = 12) +
  scale_fill_manual(values = cluster_colors)
```

#### Breeding Bird Shapefile

Finally, we need to have a GIS Shapefile that tells us the range of the breeding
birds, so that genoscape can be clipped properly.  We read this shapefile with the 
`st_read()` function from package `sf`.
```{r, results='hide'}
##Ebird range
breeding <- st_read("~/Dropbox/BGP/COYE/maps/COYE.breed_season.sf.WGS84.Ebird.shp")

ggplot(breeding) + 
  geom_sf() +
  theme_bw()

```


```{r}
library(smoothr)
area_thresh <- units::set_units(400, km^2)
breeding_dropped <- drop_crumbs(breeding, threshold = area_thresh)
# plot

br_dr<-ggplot(breeding_dropped) +
  geom_sf() +
  theme_bw() 

br_dr

??st_remove_holes
breeding_dropped2<-nngeo::st_remove_holes(breeding_dropped,4000)
br_dr2<-ggplot(breeding_dropped2,lwd=0.01) +
  geom_sf() +
  theme_bw() 

br_dr2
```

For grins, we can plot those polygons to see what they look like. But don't do this with enormous COYE shape file!
```
ggplot(breeding) + 
  geom_sf() +
  theme_bw()
```


### Preparing the data for tess3Q_map_rasters

Within Eric's fork of `tess3r` is a function called `tess3Q_map_rasters`.  It
takes input from the objects we have above, but it takes that input as 
matrices rather than data frames, etc. so there is a little finagling to be done.

#### First, make sure the lat longs are in the correct order and arrangement

We need to ensure that we have values for birds in the right order (a job for a` left_join`), and we
also have to make it a matrix with Longitude in the first column and Lat in the
second. 
```{r}
LatLong_tibble2 <-Q_tibble %>% dplyr::select(Sample) %>% left_join(met) %>% dplyr::select(Sample,Lat,Long)

long_lat_tibble <- Q_tibble %>%
  dplyr::select(Sample) %>%
  left_join(LatLong_tibble2, by = "Sample") %>%
  dplyr::select(Long, Lat) 

long_lat_matrix <- long_lat_tibble %>%
  as.matrix()
```

#### Then, make a matrix of the Q values

Pull off the names of individuals and make a matrix of it:
```{r}
Q_matrix <- Q_tibble %>%
  dplyr::select(-Sample) %>%
  as.matrix()
```

### Interpolate the Q-values by Kriging

For this, we use the above variables in `tess3r::tess3Q_map_rasters()`.  Note the
use of namespace addressing for this function rather than loading the whole `tess3r` package
with the `library()` command.
```{r}
library(tess3r)
genoscape_brick <- tess3r::tess3Q_map_rasters(
  x = Q_matrix, 
  coord = long_lat_matrix,  
  map.polygon = breeding,
  window = extent(breeding)[1:4],
  resolution = c(300,300), # if you want more cells in your raster, set higher
  # this next lines need to to be here, but don't do much...
  col.palette = tess3r::CreatePalette(cluster_colors, length(cluster_colors)), 
  method = "map.max", 
  interpol = tess3r::FieldsKrigModel(10),  
  main = "Ancestry coefficients",
  xlab = "Longitude", 
  ylab = "Latitude", 
  cex = .4
)

# after that, we need to add names of the clusters back onto this raster brick
names(genoscape_brick) <- names(Q_tibble)[-1]
summary(genoscape_brick)

writeRaster(genoscape_brick,file="COYE.genoscape_brick.raster",bylayer=F,format="raster",suffix="numbers")
```

That gives us a raster brick of Q-values associated with each cell in the raster, but
those values are not always constrained between 0 and 1, so we have to massage them 
a little bit in the next section.

### Scaling and cleaning the genoscape_brick

For this we use the function `genoscapeRtools::qprob_rando_raster()`.  This takes the raster
brick that comes out of tess3Q_map_rasters() and does some rescaling and (maybe) some random sampling
to return a raster of colors that I hope will do a reliable job of representing (in some way) predicted
assignment accuracy over space.
See `?genoscapeRtools::qprob_rando_raster` to learn about the scaling options, etc. (However, I am
not convinced that all of those options are reliably estimated.)

This will squash the raster brick down to a single
RGBA (i.e., four channels, red, green, blue and alpha) raster brick.
```{r}
genoscape_rgba <- genoscapeRtools::qprob_rando_raster(
  TRB = genoscape_brick,
  cols = cluster_colors, 
  alpha_scale = 2.0, 
  abs_thresh = 0.0, 
  alpha_exp = 1.8, 
  alpha_chop_max = 230,
  scale_min_to = 0
)

# at this point, we must be explicit about adding a projection to the raster.
# This adds the info for a regular lat-long projection
crs(genoscape_rgba) <- "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"
class(genoscape_rgba)
writeRaster(genoscape_rgba,file="COYE.genoscape_rgba.raster")
writeRaster(genoscape_rgba,file="COYE.genoscape_rgba.raster",bylayer=T,format="GTiff",suffix="numbers")
writeRaster(genoscape_rgba,file="COYE.genoscape_rgba.raster",format="GTiff",overwrite=T)

```

We can easily plot this with the function `layer_spatial` from the `ggspatial` package:
```{r}
ggplot() + 
  ggspatial::layer_spatial(genoscape_rgba) + 
  theme_bw() +
  coord_sf()
#ggsave("COYE.just_color.no_transparent.genoscape.pdf")
```

Note that if we wanted to add the actual sampling points to this (that are given
in `long_lat_tibble`), we can use ggspatial's `geom_spatial_point()` function.
```{r}
ggplot() + 
  layer_spatial(genoscape_rgba) + 
  geom_spatial_point(data = LatLong_tibble, mapping = aes(x = meanLong, y = meanLat)) + 
  #geom_text(data=LatLong_tibble, mapping = aes(x = meanLong, y = meanLat,label = State))+
  theme_bw() +
  coord_sf()
ggsave("COYE.just_color.pop_points.genoscape.colorblind.Ebird.pdf")


```

It would probably be better to jigger those points a little bit.  That could be done by mutating
each of them a random bit away.

## Coastlines, Countries, States, and Provinces

Since we will be using the Natural Earth Data Set for the raster background of our map,
we will also use the Natural Earth lines and polygons.  The Natural Earth data set is
an amazing, open-source resource for making beautiful maps.  Check it out at
[naturalearthdata.com](https://www.naturalearthdata.com/).  

### Downloading the shapefiles

For coastlines, countries, and states/provinces, you will need to download
three different shape files.  We will be working with the highest resolution
Natural Earth data sets which are the `10m` versions.  Download and unzip them
to a directory within this project called `ne_shapefiles`.  You can do that
with R like this:
```{r, eval=FALSE}
dir.create("ne_shapefiles", showWarnings = FALSE)

tmpfile <- tempfile()
downloader::download(
  url = "https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/physical/ne_10m_coastline.zip",
  dest = tmpfile
)
unzip(zipfile = tmpfile, exdir = "ne_shapefiles")
```


### Plot the Coastlines and estimate where we want to clip

First read it in:
```{r}
coastlines <- st_read("~/Dropbox/BGP/genoscape_maps/shapefiles/ne_shapefiles/ne_10m_coastline.shp")
```

Now, let's just crop out the part that we want.  This is somewhat key:
right here we will define the extent in lat-long of the region that
we want to plot:
```{r}
# note! it is important to put the elements of domain in this 
# order, because the function raster::extent() is expecting thing
# to be in this order, and doesn't parse the names of the vectors
# the way sf::st_crop() does.


domain <- c(
  xmin = -170, 
  xmax = -35,
  ymin = 3,
  ymax = 75
)

coast_cropped <- st_crop(coastlines, domain)

```

Then plot the cropped part:
```{r}
ggplot(coast_cropped) + 
  geom_sf() +
  coord_sf()
```

### All the lines, plus the genoscape

OK, that was pretty reasonable.  Now crop all the lines to `domain` and plot
them, and put the genoscape on top of that, too.
```{r}
countries_cropped <-  st_read("~/Dropbox/BGP/genoscape_maps/shapefiles/ne_shapefiles/ne_10m_admin_0_boundary_lines_land.shp") %>%
  st_crop(domain)
states_cropped <- st_read("~/Dropbox/BGP/genoscape_maps/shapefiles/ne_shapefiles/ne_10m_admin_1_states_provinces_lines.shp") %>%
  st_crop(domain)
```

Now, plot it all.  Notice we do it by just adding layers.
```{r}
mapg <- ggplot() +
  geom_sf(data = coast_cropped) +
  geom_sf(data = countries_cropped, fill = NA) +
  #geom_sf(data = states_cropped, fill = NA) +
  ggspatial::layer_spatial(genoscape_rgba) +
  theme_bw() 

# now plot it under default lat-long projection
mapg +
  coord_sf() 
```

That is looking like it should, but the projection is pretty ugly.  In the next section
we will consider projection.

## Projecting

The new version of `ggplot2` makes it super easy to project everything on the fly
by passing the projection to `coord_sf()`.   For things in North America,
it seems that a Lambert conic projection does a nice job of keeping British Columbia
and Alaska from looking way too big. We can define such a projection with a string, like
this:
```{r}
lamproj <- "+proj=lcc +lat_1=20 +lat_2=60 +lat_0=40 +lon_0=-100 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs"
```

Now, let's see how our developing map looks under such a projection.  We just define `lamproj` as 
a coordinate reference system and pass it in to `coord_sf()`:
```{r}
mapg +
  coord_sf(crs = st_crs(lamproj))
```

I would call that decidedly better.

```{r}
mapg +
  coord_sf(crs = st_crs(lamproj)) +
  #geom_spatial_point(data = LatLong_tibble, mapping = aes(x = meanLong, y = meanLat)) +
  geom_spatial_point(data = winter_birds, mapping = aes(x = longitude, y = latitude))
  
ggsave("COYE.states.pop_points.genoscape.project.Ebird.pdf")
```
## The pretty-map background

Now, all that remains is to put all of this on top of a nicely tinted map that
shows landforms and things.  Note that once we have such a layer under the rest of this
stuff, we will probably omit the coastline layer, since it gets a little dark where the
coastline is highly dissected.

### Downloading Natural Earth rasters

I am going to recommed that you get the hypsometrically tinted Natural Earth map with
water bodies and rivers on it, and you might as well get the one that has some ocean 
basin coloring too.  We will put it into a directory in the project called
`ne_rasters`.  Note, this expands to about half a gig of data.
```{r}
dir.create("ne_rasters", showWarnings = FALSE)

tmpfile <- tempfile()
downloader::download(
  url = "https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/raster/HYP_HR_SR_OB_DR.zip",
  dest = tmpfile
)
unzip(zipfile = tmpfile, exdir = "ne_rasters")
```


### Read in the raster and crop it

Reading a large raster in with the raster package does not take up much
memory, because it leaves it on disk.  So we will open a connection to
the big, massive raster using the `brick` function from the `raster` package
and then crop it:
```{r}
hypso <- brick("~/Dropbox/BGP/genoscape_maps/shapefiles/ne_rasters/HYP_HR_SR_OB_DR/HYP_HR_SR_OB_DR.tif")
hypso_cropped <- crop(hypso, extent(domain)) 
```

Now, we just add `hypso_cropped` in with `ggspatial::spatial_layer()` below all of our
other layers (dropping the coastlines), and we might make country and state lines thinner (and different sizes):

Seeing that sort of small on the screen does not really do justice to it.  You can `ggsave` it
in whatever size is desired.  For whatever the final product is going to be, you will
want to mess with the line thickness on those country and state borders...

#Add winter birds
```{r}
winter_birds <- read_delim("~/Dropbox/BGP/Marius/COYE.WinterAssignment_updated03082021.txt",delim="\t") %>% dplyr::select(PopAssignment,Long,Lat) %>%  rename(longitude=Long,latitude=Lat) %>% mutate(PopAssignment2=if_else(PopAssignment=="CA","1.CA",if_else(PopAssignment=="Midwest","2.Midwest",if_else(PopAssignment=="NewEngland","3.AtlanticCoast",if_else(PopAssignment=="West","4.West","5.Southwest"))))) %>% filter(latitude<40)

winter_birds %>% dplyr::select(PopAssignment) %>% distinct()

###Wrong color assignments - here's a janky way to deal with it
  #grp1 = "#CC0000", #red, CA
  #grp2 = "#3399FF", #blue, MW
  #grp3 = "#9933CC", #purple, AtL
  #grp4 = "#009933", #green, PNW
  #grp5 = "#FF6600")  #orange, SW

library(RColorBrewer)
warm.cols <- brewer.pal(n = 9, name = "Reds")[c(2:9)]
wint.cols<-c("#CC0000","#3399FF","#9933CC","#009933","#FF6600")
CB.color.all <- c(wint.cols,warm.cols)

```
## Clipping a rectangle out of that

While that looks nice, most people might prefer to have the
background part there in a rectangle, rather than a conic surface.  We can
do that by setting the xlim and ylim to `coord_sf()`.  But, note that
we have to do that in terms of the units that the Lambert projection is referring to,
not just in simple lat-long coordinates (i.e. we have to project those too!)

So, imagine that we want to chop out a rectangle from Haida Gwai, off the coast of
British Columbia, straight down and straight across to the right (on the image).
And we would want the bottom and right sides determined by a point in the bottom
right that has the y-value of the tip of Florida, and an x-value which is essentially
at St. John, New Brunswick.  Then, we need to determine the x and y coordinates of those
points under our Lambert conformal conic projection.  We do that by making a simple features
tibble of the lat-longs for those points and then projecting it.

```{r}
# Here are lat longs for those points
pts <- tribble(
  ~place, ~long, ~lat,
  "HaidaGwai", -186.3143526, 65.7298050,
  "FloridaTip", -80.8032237, 20.0909780, 
  "St.Johns",  -45.943453, 45.291222
)

# here we turn that into an sf object
pts_sf <- st_as_sf(pts, coords = c("long", "lat"), 
                   crs = 4326)  # 4326 is shorthand for "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"

# and now we transform it to lambert
pts_lambert <- st_transform(pts_sf, crs = lamproj)

# when we print it in a notebook, we can see that the geometry is a list
# column, but can't see the values:
pts_lambert
```

So, for you to be able to see the values we can just print it like this:
```{r}
# just note the values:
st_as_text(pts_lambert$geometry)
```

That means we want xlim = c(-2010114, 2452877) and ylim = c(-1366911, 1821869).
Let's try that:
```{r}
rectangled <- ggplot() +
  ggspatial::layer_spatial(hypso_cropped) +
  geom_sf(data = countries_cropped, fill = NA, size = 0.15) +
  geom_sf(data = states_cropped, fill = NA, size = 0.1) +
  ggspatial::layer_spatial(genoscape_rgba) +
  #geom_spatial_point(data = LatLong_tibble, mapping = aes(x = meanLong, y = meanLat)) +
   geom_spatial_point(data = winter_birds, mapping = aes(x = longitude, y = latitude, color=PopAssignment2),size=1.4,position = position_jitter(width = 1e05, height = 1e05)) +
  scale_color_manual(values = CB.color.all) +
#  geom_spatial_text_repel(data = sol, mapping = aes(x = Long, y = Lat,label=`Common Yellowthroat`, cex=0.5)) +
  theme_bw() + 
  coord_sf(
    crs = st_crs(lamproj), 
     xlim = c(-3068118, 3971262), 
    ylim = c(-3099797, 3464362),
    expand = FALSE)  # expand = FALSE means put the axes right at the xlim and ylim

  #geom_spatial_point(data = LatLong_tibble, mapping = aes(x = meanLong, y = meanLat)) +
  #geom_spatial_text(data=LatLong_tibble, mapping = aes(x = meanLong, y = meanLat,label = State)) +

rectangled
```


```{r}
ggsave("COYE.genoscape.complete.wWinter_assign.clip.eBird.pdf")
```

Okay, that does it.  In reality, you would probably want to expand the original `domain`
to be a bigger chunk of the earth, so that when you cut stuff off there was more left...

##Let's add wintering birds AND ecoregions
```{r}
ecoregions<-st_read("~/Dropbox/BGP/Network_modeling/Winter_ecoregions/COYE.winterEcoregionsFinal8_all.cut.sample.sf.WGS84.Ebird.shp")
ecoregions.sub <- readOGR(dsn = "~/Dropbox/BGP/Network_modeling/Winter_ecoregions/",
          layer="COYE.winterEcoregionsFinal8_all.cut.sample.sf.WGS84.Ebird")

ecoregions.sub.simplified <- rgeos::gSimplify(ecoregions.sub,
                                              tol = 0.05,
                                              topologyPreserve = T)
summary(ecoregions.sub.simplified)
ecoregions.sub.spdf <- SpatialPolygonsDataFrame(ecoregions.sub.simplified, ecoregions.sub@data)
ecoregions.sub.sf <- st_as_sf(ecoregions.sub.spdf)
```

```{r}
winter_birds <- read_delim("~/Dropbox/BGP/Marius/COYE.WinterAssignment_updated03082021.txt",delim="\t") %>% dplyr::select(PopAssignment,Long,Lat) %>%  rename(longitude=Long,latitude=Lat) %>% mutate(PopAssignment2=if_else(PopAssignment=="CA","1.CA",if_else(PopAssignment=="Midwest","2.Midwest",if_else(PopAssignment=="NewEngland","3.AtlanticCoast",if_else(PopAssignment=="West","4.West","5.Southwest"))))) %>% filter(latitude<40)

winter_birds %>% dplyr::select(PopAssignment) %>% distinct()

###Wrong color assignments - here's a janky way to deal with it
  #grp1 = "#CC0000", #red, CA
  #grp2 = "#3399FF", #blue, MW
  #grp3 = "#9933CC", #purple, AtL
  #grp4 = "#009933", #green, PNW
  #grp5 = "#FF6600")  #orange, SW

library(RColorBrewer)
warm.cols <- brewer.pal(n = 9, name = "Greys")[c(1:9)]
wint.cols<-c("#CC0000","#3399FF","#9933CC","#009933","#FF6600")
CB.color.all <- c(wint.cols,warm.cols)

##genoscape size
#     xlim = c(-2868118, 3571262), 
#    ylim = c(-2799797, 3864362),

#Just winter ecoregions and assignment  
rectangled2 <- ggplot() +
  geom_sf(data = coast_cropped) +
  geom_sf(data = countries_cropped, fill = NA) +
  #geom_sf(data = states_cropped, fill = NA)   geom_sf(data = countries_cropped, fill = NA, size = 0.15) +
  geom_sf(data = states_cropped, fill = NA, size = 0.1) +
  #ggspatial::layer_spatial(genoscape_rgba) +
  #geom_spatial_point(data = LatLong_tibble, mapping = aes(x = meanLong, y = meanLat)) +
   geom_sf(data = ecoregions,mapping = aes(color=Region2),lwd=0.1) +
  geom_spatial_point(data = winter_birds, mapping = aes(x = longitude, y = latitude, color=PopAssignment2),size=1.6,position = position_jitter(width = 1e05, height = 9e04)) +
  scale_color_manual(values = CB.color.all) +
  theme_bw() + 
  coord_sf(
    crs = st_crs(lamproj), 
    xlim = c(-2868118, 3571262), 
    ylim = c(-2799797, 294362),
    expand = FALSE) +xlab("Longitude") +ylab("Latitude")
rectangled2
```

```{r}
ggsave("COYE.ecoregion.wWint_assign.jitter.Grey.pdf")
```

```
# now, let's put the migrants on there in color:


```

## A final note

If you want to add more elements to the figure you can do so by giving them
a lat-long and making a simple features tibble out of it and then putting them
there using geom_sf().  It all fits nicely into the ggplot framework.

## Session Info

```{r}
sessioninfo::session_info()
```